### DISCUSSION TITLE
Many Windows objects spy for accessibility/automation IDs only show top level information

### DISCUSSION THREAD

**Discussion by djalan** [AvaloniaUI/Avalonia#15571] [General]
I would like to know if sub-level information is availabe for non-headless automation. I used many objects spy, but unfortunately they are not able to find all the objects.

When spying Calculator, you can see many groups of groups containing different objects.

![flaui_calculator](https://github.com/AvaloniaUI/Avalonia/assets/707389/c4b43943-0262-48e9-981d-db63d2da5f46)

But for our Avalonia App, it seems that it stops at ListItem and we can't see the TextBlock under ListItem and there is no way to retrieve its text value.

![flaui_carousel](https://github.com/AvaloniaUI/Avalonia/assets/707389/60fb3ebe-03fd-4f98-ba44-d2f2c78d675b)

We nested many levels of .axaml files for modular development. Is this breaking the visibility of objects when using a spy?

**Comment by stevemonaco** [comment]
Can you provide a repro or XAML sample that demonstrates? I have scenarios that populate `ListItem` and I have scenarios that do not (viewed within FlaUIInspect). The type of child controls used is important. In the latter case, I had `Border`s and a `Rectangle` for a palletized color picker. Neither of those implement a real `AutomationPeer` and uses the default `NoneAutomationPeer`, I believe.

**Reply by djalan** [reply]
We are using a customized TextBlock that changes size automatically named ResizableTextBlock.

I was able to override the below method

https://github.com/AvaloniaUI/Avalonia/blob/27e446e5e06c9fd35b6d1fd59cd1dab6445ed359/src/Avalonia.Controls/TextBlock.cs#L773

and made a copy of the below that we called ResizableTextBlockAutomationPeer

https://github.com/AvaloniaUI/Avalonia/blob/27e446e5e06c9fd35b6d1fd59cd1dab6445ed359/src/Avalonia.Controls/Automation/Peers/TextBlockAutomationPeer.cs#L5

And now two different object spys can see my object under ListItem.

**Reply by stevemonaco** [reply]
If you derive from `TextBlock`, then you should be fine with the parent's setting of `TextBlockAutomationPeer`. At least my derived `TextBlock` shows up...though it's not in a collection control.

If it works, it works I guess. I don't know the AutomationPeer area well.

**Reply by djalan** [reply]
It used to be derived from `TextBlock`. Someone changed the code and it is now derived from `TemplatedControl`.

**Reply by stevemonaco** [reply]
That would do it. Without overriding, you would be exposing `NoneAutomationPeer` since that's the default.
